### ISSUE TITLE
SHA 256

### ISSUE THREAD

**Issue by vbjay** [gitextensions/gitextensions#5483]
**Do you want to request a *feature* or report a *bug*?**
Feature
**What is the current behavior?**
We only support sha1.
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

**What is the expected behavior?**
Supporting sha 256 also as mentioned here. https://blog.github.com/2018-09-10-highlights-from-git-2-19/#transitioning-away-from-sha-1 
**Environment you encounter the issue:**

- GitExtensions version:Master
- GIT version: 2.19 +
- OS version: NA
- .NET version: NA

**Did this work in previous version of GitExtensions (which)?**
no

**Comment by gerhardol** [comment]
GE is being prepared with the ObjectId abstraction (also for memory and speed reasons).
But I see no need to support SHA256 officially yet.
I suggest that this is closed for now.

**Comment by drewnoakes** [comment]
I believe Git now supports arbitrary hashes, so any future work here should be thinking beyond SHA256.

**Comment by vbjay** [comment]
Correct. I just wanted everyone to be on the same page and we would need to
alter our code to support the different hashes and their output.

**Comment by Chri-s** [comment]
Just wanted to give some feedback since git officially supports sha256 hashes now.

I just tried to open a git repo with sha256 hashes (created with `git init --object-format=sha256`) in Git extensions.
GE shows the commits, but all information except the commit message is wrong.

For the commit

```
$ git log
commit 82eddb6a8066867882c3e0caa3f4ff1d43120c48c956159b9bd3feb6f5d514e3 (HEAD -> master)
Author: chri-s <chri-s@test.com>
Date:   Sat Apr 20 22:19:22 2024 +0200

    Add scripts and README
```

GE shows
```
Author: 438f6a3ef46fdae367f2ea5db809104376197f694dd40c448635de69cd05553e542eec8b77ab0d3c1c003290f405000b912ae531 <1713644362>
Author date: 31.12.9999 23:59:59
Committer: 1713644362 <chri-s>
Commit date: 01.01.1970 02:46:04
Commit hash: 82eddb6a8066867882c3e0caa3f4ff1d43120c48

Add scripts and README
```

The branches are not recognized in the left tree view.
If I click on the Diff tab in a commit, this exception is raised:

- Git Extensions 4.2.1.17611
- Build b0c0b284876b63c52c77195ada82a85bfadbb083
- Git 2.44.0.windows.1
- Microsoft Windows NT 10.0.22631.0
- .NET 6.0.29
- DPI 96dpi (no scaling)
- Portable: False

Exit code: 128
Command: C:\Program Files\Git\bin\git.exe
Arguments: ls-tree -z -r c956159b9bd3feb6f5d514e3c2c02eb9506f95d5
Working directory: D:\Development\svn2gitrules-sha256\

```
GitExtUtils.ExternalOperationException: fatal: Not a valid object name c956159b9bd3feb6f5d514e3c2c02eb9506f95d5
 ---> System.Exception: fatal: Not a valid object name c956159b9bd3feb6f5d514e3c2c02eb9506f95d5
   --- End of inner exception stack trace ---
   at async Task<string> GitCommands.ExecutableExtensions.GetOutputAsync(IExecutable executable, ArgumentString arguments, byte[] input, Encoding outputEncoding, CommandCache cache, bool stripAnsiEscapeCodes) in C:/projects/gitextensions/GitCommands/Git/ExecutableExtensions.cs:line 129
   at void Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()
   at T Microsoft.VisualStudio.Threading.JoinableTask<T>.CompleteOnCurrentThread()
   at T Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run<T>(Func<Task<T>> asyncMethod, JoinableTaskCreationOptions creationOptions) x 2
   at string GitCommands.ExecutableExtensions.GetOutput(IExecutable executable, ArgumentString arguments, byte[] input, Encoding outputEncoding, CommandCache cache, bool stripAnsiEscapeCodes) in C:/projects/gitextensions/GitCommands/Git/ExecutableExtensions.cs:line 43
   at IEnumerable<INamedGitItem> GitCommands.GitModule.GetTree(ObjectId commitId, bool full) in C:/projects/gitextensions/GitCommands/Git/GitModule.cs:line 3397
   at IReadOnlyList<GitItemStatus> GitCommands.GitModule.GetTreeFiles(ObjectId commitId, bool full) in C:/projects/gitextensions/GitCommands/Git/GitModule.cs:line 2719
   at IReadOnlyList<FileStatusWithDescription> GitUI.FileStatusDiffCalculator.SetDiffs(IReadOnlyList<GitRevision> revisions, ObjectId headId, bool allowMultiDiff, CancellationToken cancellationToken) in C:/projects/gitextensions/GitUI/UserControls/FileStatusDiffCalculator.cs:line 51
   at async Task GitUI.FileStatusList.SetDiffsAsync(IReadOnlyList<GitRevision> revisions, ObjectId headId, CancellationToken cancellationToken) in C:/projects/gitextensions/GitUI/UserControls/FileStatusList.cs:line 765
   at async Task GitUI.CommandsDialogs.RevisionDiffControl.SetDiffsAsync(IReadOnlyList<GitRevision> revisions) in C:/projects/gitextensions/GitUI/CommandsDialogs/RevisionDiffControl.cs:line 280
   at async void GitUI.CommandsDialogs.RevisionDiffControl.DisplayDiffTab(IReadOnlyList<GitRevision> revisions)+(?) => { } in C:/projects/gitextensions/GitUI/CommandsDialogs/RevisionDiffControl.cs:line 244
   at async void GitUI.TaskManager.InvokeAndForget(Control control, Func<Task> asyncAction, CancellationToken cancellationToken)+(?) => { } [1] in C:/projects/gitextensions/GitExtUtils/GitUI/TaskManager.cs:line 111
   at async Task GitUI.TaskManager.HandleExceptionsAsync(Func<Task> asyncAction, Func<Exception, Task> handleExceptionAsync) in C:/projects/gitextensions/GitExtUtils/GitUI/TaskManager.cs:line 38
```

**Comment by gerhardol** [comment]
GE internally stores the object id as sha1, so sha256 will not work. I did not expect any info to be shown...
The app is somehow prepared for changing storage, but it is considerable work still to support sha256.

**Comment by Chri-s** [comment]
I made some changes in the draft PR #11707, basically changing the ObjectId to be able to hold a sha1 or sha256 hash and changing some regexes. I also added dots to the git log format so that RevisionReader can see where a hash ends.

With these changes, GE can parse and display the git log of a sha256 repository, displays the branches and the commit tabs such as Files or Diff also work.

Can you or someone else check whether this is a sensible approach for the project? Then I would try to continue in this way.
I can't promise to complete the changes without help because I don't have detailed knowledge of git internals.

**Comment by gerhardol** [comment]
Can you read the configuration for a repo and get the length of the sha?
The parsing of sha is time critical when loading a repo, to search for the length of the sha for every sha is slowing down for all.

**Comment by pmiossec** [comment]
FYI, from what i understand from this article ( https://www.phoronix.com/news/Git-2.46-rc0-Released ), next version will start to initialize new repos with sha256...

If that's true, I think we will see error reports of users not able to open their repository with gitextension.

**Comment by gerhardol** [comment]
> FYI, from what i understand from this article ( https://www.phoronix.com/news/Git-2.46-rc0-Released ), next version will start to initialize new repos with sha256...

For Git 3.0. No time plan what I see, other than "not soonish".

**Comment by pmiossec** [comment]
Reading it again, you're right.

I was mistaken by: 

>Git 2.46-rc0 is also no longer setting SHA1 as the default object hash type when initializing a repository.

But it seems that behavior is still kept the same by something I dont really understand:

>That initial setup for SHA1 ultimately gets overrode by most Git commands during the directory setup process,

**Comment by gerhardol** [comment]
Draft implementation in #11707

**Comment by pmiossec** [comment]
From Git project, it seems that release date  is planned around the end of 2026 : https://www.phoronix.com/news/Git-3.0-Release-Talk-2026
